### PR TITLE
Enhancement to Message classes

### PR DESCRIPTION
**Pull request by Eduardo-T**

### CODE REVIEW

File: src/main/java/ai/knowly/langtoch/capability/modality/text/ChatCompletionLLMCapability.java [hunk 1]
```diff
@@ -7,73 +7,69 @@
...
-      Optional<Parser<I, MultiChatMessage>> inputParser,
-      Optional<Parser<ChatMessage, O>> outputParser) {
+          Processor<MultiChatMessage, ChatMessage> processor,
+          Parser<I, MultiChatMessage> inputParser,
+          Parser<ChatMessage, O> outputParser) {
     this.processor = processor;
     this.inputParser = inputParser;
     this.outputParser = outputParser;
-    this.memory = Optional.empty();
+    this.memory = null;
```
Comment (li2109) [inline src/main/java/ai/knowly/langtoch/capability/modality/text/ChatCompletionLLMCapability.java:29]: Let's revert this change and not putting null here which is error-prone.

Comment (li2109) [reply src/main/java/ai/knowly/langtoch/capability/modality/text/ChatCompletionLLMCapability.java:29]: We could write more helper functions to make it easier to use

File: src/main/java/ai/knowly/langtoch/schema/chat/ChatMessage.java
Comment (li2109) [inline src/main/java/ai/knowly/langtoch/schema/chat/ChatMessage.java:1]: Please write a test for this.

**Review by li2109 — CHANGES_REQUESTED** [review]
Everything looks good for the Message class.

I would rather against removing the optional

File: src/main/java/ai/knowly/langtoch/capability/modality/text/ChatCompletionLLMCapability.java (hunk 1 above)
Comment (Eduardo-T) [reply src/main/java/ai/knowly/langtoch/capability/modality/text/ChatCompletionLLMCapability.java:29]: > Let's revert this change and not putting null here which is error-prone.

I don't think is error-prone. Optionals should be used as a method return type, not as a field.

Comment (li2109) [reply src/main/java/ai/knowly/langtoch/capability/modality/text/ChatCompletionLLMCapability.java:29]: Hah, it's more like a design philosophy here.

Personally, My philosophy aligns with the idea: Null value is the billion dollar mistake. https://www.infoq.com/presentations/Null-References-The-Billion-Dollar-Mistake-Tony-Hoare/

I don't like null. Wrapping Optional on top of variables that could potentially undefined making null check explicit. the biggest advantage of Optional is its idiot-proof-ness

I understand that JDK's Optional is unserializable as they did this for purpose. 

However, Google's guava Optional is serializable suggesting Optional can be used as a field.
You can see Optional as field usage in google's library:   https://github.com/google/tink/blob/9bc2667963e20eb42611b7581e570f0dddf65a2b/java_src/src/main/java/com/google/crypto/tink/jwt/JwtHmacKeyManager.java#L101

Comment (Eduardo-T) [reply src/main/java/ai/knowly/langtoch/capability/modality/text/ChatCompletionLLMCapability.java:29]: > Hah, it's more like a design philosophy here.
> 
> Personally, My philosophy aligns with the idea: Null value is the billion dollar mistake. https://www.infoq.com/presentations/Null-References-The-Billion-Dollar-Mistake-Tony-Hoare/
> 
> I don't like null. Wrapping Optional on top of potentially undefined variables makes null check explicit. the biggest advantage of Optional is its idiot-proof-ness
> 
> I understand that JDK's Optional is unserializable as they did this for purpose.
> 
> However, Google's guava Optional is serializable suggesting Optional can be used as a field. From this, we can peek Google's design philosophy which I would prefer over JDK's. You can see Optional as field usage in many google's library: On example: https://github.com/google/tink/blob/9bc2667963e20eb42611b7581e570f0dddf65a2b/java_src/src/main/java/com/google/crypto/tink/jwt/JwtHmacKeyManager.java#L101

I agree with you on the null dangers and indiscriminate use of it, but I kindly disagree with you on the way Optionals have been used in this part of the code. (Maybe in other parts too)

For what I've seen, Optionals are being accepted as parameters. That is a redundant fault. Imagine passing by accident a null Optional. The optional now lost its purpose.

The idea behind Optional is exactly prevent that from happening, technically, Optionals should be used on data retrieval, like you did with the getMemory() I guess.

getSomething().ifPresent(o -> perform::something);
getSomething().orElse(defaultValue).performSomething();

From https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Optional.html
"Optional is primarily intended for use as a method return type where there is a clear need to represent "no result," and where using null is likely to cause errors. A variable whose type is Optional should never itself be null; it should always point to an Optional instance."

Comment (li2109) [reply src/main/java/ai/knowly/langtoch/capability/modality/text/ChatCompletionLLMCapability.java:29]: Sorry about the late reply. I was too busy these two days to resolve the comments.

Yes, it's actually a debatable topic and I used to work for two companies that have opposite stances on Optional, just like this issue.  
the JDK style actually initiated intense debate: https://mail.openjdk.org/pipermail/jdk8-dev/2013-September/003186.html 

It's not right or wrong, but more like preference and purely depends on which style you are following and just being consistent. like in Python, 4 spaces vs 2 spaces. either is fine and just don't mix it.

Sorry, I think I am gonna "be consistent" 😀

i am going to add a commit to only add the enhancement of message class and merge to the master

Comment (yeefom) [reply src/main/java/ai/knowly/langtoch/capability/modality/text/ChatCompletionLLMCapability.java]: Randomly saw this super interesting conversation! Wanted to offer my unsolicited 2 cents as I apparently have too much time. 😃 

I agree with @Eduardo-T for reserving the usage of Optional as return type to indicate intention — this method _will_ return null in some cases. I see fields a closer concept to variables which are acceptable to be null upon initiation. In other ways, a field _could_ be null. 

In Java there will always be null. The cost of wrapping every nullable field with Optional might be higher than developer speed — for one, some Optionals might inevitably be passed as variables around.
